docs/core/diagnostics/debug-deadlock.md

```diff
@@ -0,0 +1,287 @@
+- [dotnet-trace](dotnet-trace.md) to list processes
+- [dotnet-dump](dotnet-dump.md) to collect, and analyze a dump file
+
+## Core dump generation
+
+To investigate application hang, a core dump (memory dump) allows you to inspect the state of its threads, and any possible locks that may have contention issues. Run the [sample debug](https://docs.microsoft.com/samples/dotnet/samples/diagnostic-scenarios) application using the following command from the sample root directory.
+
+```dotnetcli
+dotnet run
+```
+
+To find the process ID, use the following command:
+
+```dotnetcli
+dotnet-trace ps
+```
+
+Take note of the process ID from your command output (yours will be different), ours was `4807`. Navigate to the following URL, which is an API endpoint on the sample site:
+
+[https://localhost:5001/api/diagscenario/deadlock](https://localhost:5001/api/diagscenario/deadlock)
+
+The API request to the site will hang, and not respond - let the request run for about 10-15 seconds. Then create the core dump using the following command:
+
+### [Linux](#tab/linux)
+
+```bash
+sudo dotnet-dump collect -p 4807
+```
+
+### [Windows](#tab/windows)
+
+```console
+dotnet-dump collect -p 4807
+```
+
+---
+
+## Analyzing the core dump
+
+To start the core dump analysis, open the core dump using the following `dotnet-dump analyze` command. The argument is the path to the core dump file that was collected earlier.
+
+```dotnetcli
+dotnet-dump analyze  ~/.dotnet/tools/core_20190513_143916
+```
+
+Since you're looking at a potential hang, you want an overall feel for the thread activity in the process. You can use the `threads` command as shown below:
+
+```console
+> threads
+*0 0x1DBFF (121855)
+ 1 0x1DC01 (121857)
+ 2 0x1DC02 (121858)
+ 3 0x1DC03 (121859)
+ 4 0x1DC04 (121860)
+ 5 0x1DC05 (121861)
+ 6 0x1DC06 (121862)
+ 7 0x1DC07 (121863)
+ 8 0x1DC08 (121864)
+ 9 0x1DC09 (121865)
+ 10 0x1DC0A (121866)
+ 11 0x1DC0D (121869)
+ 12 0x1DC0E (121870)
+ 13 0x1DC10 (121872)
+ 14 0x1DC11 (121873)
+ 15 0x1DC12 (121874)
+ 16 0x1DC13 (121875)
+ 17 0x1DC14 (121876)
+ 18 0x1DC15 (121877)
+ 19 0x1DC1C (121884)
+ 20 0x1DC1D (121885)
+ 21 0x1DC1E (121886)
+ 22 0x1DC21 (121889)
+ 23 0x1DC22 (121890)
+ 24 0x1DC23 (121891)
+ 25 0x1DC24 (121892)
+ 26 0x1DC25 (121893)
+...
+...
+ 317 0x1DD48 (122184)
+ 318 0x1DD49 (122185)
+ 319 0x1DD4A (122186)
+ 320 0x1DD4B (122187)
+ 321 0x1DD4C (122188)
+ ```
+
+The output shows all the threads currently running in the process, with their associated debugger thread ID, and operating system thread ID. Based on the output, there are little over 300 threads.
+
+The next step is to get a better understanding of what the threads are currently doing by getting each thread's callstack. The `clrstack` command can be used to output callstacks. It can either output a single callstack, or all the callstacks. Use the following command to output all the callstacks for all the threads in the process.
```

Comment:
```suggestion
The next step is to get a better understanding of what the threads are currently doing by getting each thread's callstack. The `clrstack` command can be used to output callstacks. It can either output a single callstack or all the callstacks. Use the following command to output all the callstacks for all the threads in the process:
```